rust-test-program/src/hello_world.rs

```diff
@@ -0,0 +1,68 @@
+#![no_std]
+#![no_main]
+
+use core::panic::PanicInfo;
+use core::ptr::write_volatile;
+
+/// Panic handler for bare metal - infinite loop on panic
+#[panic_handler]
+fn panic(_info: &PanicInfo) -> ! {
+    loop {}
+}
+
+/// Write to tohost address to signal completion
+#[inline(never)]
+fn write_tohost(value: u32) -> ! {
+    const TOHOST_ADDR: u32 = 0xFFFF_FFF0;
+    unsafe {
+        write_volatile(TOHOST_ADDR as *mut u32, value);
+    }
+    loop {}
+}
+
+/// FIFO memory-mapped I/O addresses
+const FIFO_BASE: u32 = 0x4000_0000;
+const FIFO_DATA: u32 = FIFO_BASE + 0x0;
+const FIFO_STATUS: u32 = FIFO_BASE + 0x4;
+
+/// FIFO status bits
+const _FIFO_RX_VALID: u32 = 0x1;
+const FIFO_TX_READY: u32 = 0x2;
+
+/// Write a byte to the FIFO
+#[inline(never)]
+fn fifo_write_byte(byte: u8) {
+    unsafe {
+        // Wait for FIFO to be ready
+        loop {
+            let status = core::ptr::read_volatile(FIFO_STATUS as *const u32);
+            if (status & FIFO_TX_READY) != 0 {
+                break;
+            }
+        }
+
+        // Write byte to FIFO
+        write_volatile(FIFO_DATA as *mut u32, byte as u32);
```

Comment:
It looks like this code will write a whole word instead of a single byte which doesn't seem correct. The hardware fifo can only be written or read in u32 increments so we need new logic in this file that chunks up strings into u32s before writing them into the fifo. They'll also need to be zero padded at the end in the case where they don't align to 4 bytes.